eslint-local-rules.js

```diff
@@ -33,4 +33,42 @@ module.exports = {
       }
     },
   },
+  'no-async-module-init': {
+    meta: {
+      type: 'problem',
+      docs: {
+        description: 'disallow async module initialisation',
+        category: 'Possible Errors',
+        recommended: true,
+        url: '',
+      },
+      messages: {
+        noAsyncRegister:
+          'Disallowing async register functions in modules to prevent unexpected startup failures or timeouts.',
+        noAsyncProviderFactory:
+          'Disallowing async useFactory in module providers to prevent unexpected startup failures or timeouts.',
+      },
+    },
+    schema: [],
+    create: function (context) {
+      return {
+        "MethodDefinition[static=true][key.name='register'][value.async=true]": (
```

Comment:
Did not know you could implement an eslint rule like this. It's really cool!

I wonder about the selectors though. This one might have false positives (async register functions not related to Nest) as well as false negatives. "register" is just a naming convention, "forRootAsync" is another common name for async module imports. 

Maybe we should select for what they return instead: `Promise<DynamicModule>` - though that depends on the user specifying the correct return type.

Your call. Maybe report all three selectors for now and loosen up if it's too annoying?